### PR TITLE
Imagenet Classify Example

### PR DESCRIPTION
**Pull request by mariusbld**
Using an ImageNet model to do inference with our `model-servers`. Documentation will follow in our main repo, linking to these files.

The main purpose of this python script is to make it more out-of-the-box user friendly to experience `model-servers`. We're passing in an image URL and printing out the English words corresponding to the detected object.

### CODE REVIEW

File: imagenet/classify_resnet.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    print('Classify result: {}'.format(cls_name))
+
+
+if __name__ == '__main__':
+    # Ignore warnings due to SSL certificate for prediction server REST api.
+    if not sys.warnoptions:
+        warnings.simplefilter("ignore")
+
+    parser = argparse.ArgumentParser(description="Classify image using an ImageNet model.")
+    parser.add_argument('--img', default=IMAGE_URL, help='JPEG image to classify. Must be an URL path.')
```
Comment (TreyLawrence) [inline imagenet/classify_resnet.py]: why only url?

File: imagenet/classify_resnet.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import argparse
+import base64
+import json
+import requests
+import sys
+import warnings
+
+
+IMAGE_URL = 'https://github.com/spellrun/modelservers/raw/master/imagenet/image.jpg'
+SERVER_URL = 'https://serving.spell.run/marius/resnet/v1/rest/model:predict'
```
Comment (TreyLawrence) [inline imagenet/classify_resnet.py]: can you move it to be on the spellrun organization?

File: imagenet/classify_resnet.py (hunk 1 above)
Comment (mariusbld) [reply imagenet/classify_resnet.py]: I thought it would be easier for them to run the script without downloading stuff on their machine. Should we switch only to local file, or implement both?

File: imagenet/classify_resnet.py (hunk 2 above)
Comment (mariusbld) [reply imagenet/classify_resnet.py]: I'll talk to @jayschwa about this, I think we need to be part of the org in prod as well.

File: imagenet/classify_resnet.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    res = requests.get(url, stream=True)
+    res.raise_for_status()
+    return res.content
+
+
+def predict_img_class(img_bytes, server, auth):
+    print('Predicting image class..')
+    data = '{"instances" : [{"b64": "%s"}]}' % base64.b64encode(img_bytes)
+    headers = {'Authorization':'Bearer {}'.format(auth)}
+    res = requests.post(server, data=data, verify=False, headers=headers)
```
Comment (bgardiner) [inline imagenet/classify_resnet.py]: why `verify=False`? we have a valid SSL cert for serving.spell.run.

File: imagenet/classify_resnet.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    prediction = res.json()['predictions'][0]
```
Comment (bgardiner) [inline imagenet/classify_resnet.py]: add `res.raise_for_status()`?

File: imagenet/classify_resnet.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    img_bytes = load_jpeg_img(img)
+    img_cls = predict_img_class(img_bytes, server, auth)
+    cls_name = img_class_to_name(img_cls)
+    print('Classify result: {}'.format(cls_name))
+
+
+if __name__ == '__main__':
+    # Ignore warnings due to SSL certificate for prediction server REST api.
+    if not sys.warnoptions:
+        warnings.simplefilter("ignore")
```
Comment (bgardiner) [inline imagenet/classify_resnet.py]: why is this needed? I think it's bad practice to be suppressing SSL warnings for code we're asking people to run

File: imagenet/classify_resnet.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    headers = {'Authorization':'Bearer {}'.format(auth)}
+    res = requests.post(server, data=data, verify=False, headers=headers)
+    prediction = res.json()['predictions'][0]
+    return prediction['classes']
+
+
+def img_class_to_name(img_class_id):
+    json_data=open(IMAGE_NET_CLS).read()
+    data = json.loads(json_data)
+    return data[str(img_class_id - 1)]
```
Comment (bgardiner) [inline imagenet/classify_resnet.py]: why `img_class_id - 1`? the image classes are 1-indexed?

Comment (bgardiner) [reply imagenet/classify_resnet.py]: No i think it’s fine (just a little weird) . Mostly just wanted to make sure I was tracking

File: imagenet/classify_resnet.py (hunk 1 above)
Comment (mariusbld) [reply imagenet/classify_resnet.py]: I gave up on the idea to make it work for both, it brings in a weird library `urlparse` to detect if the path is an URL. The library is different for python2 vs python3 (`urlib.parse` vs `urlparse`) and I wanted to keep this example as simple as possible. I could switch to local JPG file, but then the script won't just work out-of-the-box with the default image that we provide.

Comment (bgardiner) [reply imagenet/classify_resnet.py]: you could use the `six.moves.urllib`.... it is compatible with both python2 and python3. That is what we use elsewhere in the spell python code, so it should already be installed if spell is installed

File: imagenet/classify_resnet.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def load_jpeg_img(path):
+    print('Loading JPEG image: {}'.format(path))
+    if is_url(path):
+        res = requests.get(path, stream=True)
+        res.raise_for_status()
+        return res.content
+    else:
+        return open(path, "rb").read()
```
Comment (bgardiner) [inline imagenet/classify_resnet.py]: nit: ensure the file handle is closed
```suggestion
        with open(path, "rb") as f:
            return f.read()
```
